File: app14_LUA_UNITY/backend/spec/routes/auth_spec.lua
require("spec.spec_helper")
local http = require("spec.support.http_client")
local JwtService = require("app.services.jwt_service")

describe("POST /api/v1/auth/login", function()
  it("returns a bearer token for the correct credentials (D-01)", function()
    local res = http.post("/api/v1/auth/login", { username = TEST_ADMIN_USERNAME, password = TEST_ADMIN_PASSWORD })

    assert.are.equal(200, res.status)
    assert.are.equal("Bearer", res.json.tokenType)
    assert.are.equal("ADMIN", res.json.role)
    assert.is_true(#res.json.token > 0)

    local decoded = JwtService.decode(res.json.token)
    assert.are.equal(TEST_ADMIN_USERNAME, decoded.sub)
    assert.are.equal("ADMIN", decoded.role)
  end)

  it("401s for a wrong password", function()
    local res = http.post("/api/v1/auth/login", { username = TEST_ADMIN_USERNAME, password = "definitely-wrong" })
    assert.are.equal(401, res.status)
  end)

  it("401s for an unknown username", function()
    local res = http.post("/api/v1/auth/login", { username = "does-not-exist", password = "whatever" })
    assert.are.equal(401, res.status)
  end)

  it("400s when username is missing", function()
    local res = http.post("/api/v1/auth/login", { password = "whatever" })
    assert.are.equal(400, res.status)
  end)

  -- SR-05: throttled per source IP after 5 attempts within 60 seconds.
  it("rate-limits repeated login attempts from the same IP", function()
    local res
    for _ = 1, 6 do
      res = http.post("/api/v1/auth/login", { username = TEST_ADMIN_USERNAME, password = "wrong" })
    end
    assert.are.equal(429, res.status)
  end)
end)

File: app13_ruby_FastApi/backend/spec/requests/auth_spec.rb
# frozen_string_literal: true

require_relative "../spec_helper"

RSpec.describe "POST /api/v1/auth/login" do
  it "returns a bearer token for the correct credentials (D-01)" do
    post "/api/v1/auth/login", { username: TEST_ADMIN_USERNAME, password: TEST_ADMIN_PASSWORD }.to_json,
         "CONTENT_TYPE" => "application/json"

    expect(last_response.status).to eq(200)
    body = JSON.parse(last_response.body)
    expect(body["tokenType"]).to eq("Bearer")
    expect(body["role"]).to eq("ADMIN")
    expect(body["token"]).not_to be_empty

    decoded = JwtService.decode(body["token"])
    expect(decoded["sub"]).to eq(TEST_ADMIN_USERNAME)
    expect(decoded["role"]).to eq("ADMIN")
  end

  it "401s for a wrong password" do
    post "/api/v1/auth/login", { username: TEST_ADMIN_USERNAME, password: "definitely-wrong" }.to_json,
         "CONTENT_TYPE" => "application/json"
    expect(last_response.status).to eq(401)
  end

  it "401s for an unknown username" do
    post "/api/v1/auth/login", { username: "does-not-exist", password: "whatever" }.to_json,
         "CONTENT_TYPE" => "application/json"
    expect(last_response.status).to eq(401)
  end

  it "400s when username is missing (D-02: Grape params validation)" do
    post "/api/v1/auth/login", { password: "whatever" }.to_json, "CONTENT_TYPE" => "application/json"
    expect(last_response.status).to eq(400)
  end

  # SR-05: throttled per source IP after 5 attempts within 60 seconds.
  it "rate-limits repeated login attempts from the same IP" do
    6.times do
      post "/api/v1/auth/login", { username: TEST_ADMIN_USERNAME, password: "wrong" }.to_json,
           "CONTENT_TYPE" => "application/json"
    end
    expect(last_response.status).to eq(429)
  end
end
